go/pkg/rename/rename-netspoc.1.md
# rename-netspoc 1 "" Netspoc "User Manual"

# NAME

rename-netspoc - Rename one or more objects in netspoc files

# SYNOPSIS

rename-netspoc [options] FILE|DIR SUBSTITUTION ...

# DESCRIPTION

This program reads a netspoc configuration and one or more
SUBSTITUTIONS. It substitutes found objects with its replacement in
each file. Changes are done in place, no backup files are created. But
only changed files are touched.

## SUBSTITUTION

A SUBSTITUTION is a pair of typed names "type:NAME1 type:NAME2".
NAME1 is searched and replaced by NAME2. Both types of a single
SUBSTITUTION must use the same type.
Multiple SUBSTITUTIONS can be applied in a single run of rename-netspoc.

A typed name can use any valid type and name in netspoc syntax.
Valid types are: `router network host any group area service owner
protocol protocolgroup pathrestriction nat isakmp ipsec crypto`.

A SUBSTITUTION of type `network` also changes `interface`s and `host:id`
which reference the given network.

A SUBSTITUTION of type `router` also changes `interface`s
which reference the given router.

A SUBSTITUTION of type `nat` also changes the corresponding
`nat_in` and `nat_out` lists.

# OPTIONS

**-f**, **--file** file
:   Read SUBSTITUTIONS from file.

**-q**, **--quiet**
:   Don't print status messages.

**-h**, **--help**
:   Print a brief help message and exit.

# EXAMPLES

Rename `network:x` to `network:y` of Netspoc configuration in
directory `netspoc/`.
This also changes interface names referencing `network:x`,
e.g. `interface:r.x` is renamed to `interface:r.y` .

`rename-netspoc netspoc/ network:x network:y`

# COPYRIGHT AND DISCLAIMER

(c) 2025 by Heinz Knutzen, heinz.knutzen@googlemail.com

This program is part of Netspoc, a Network Security Policy Compiler.
http://hknutzen.github.com/Netspoc

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


Changes
{{$NEXT}}

 - export-netvis: export Dualstack-Address of network
   in field Address2 of node.

6.079     2025-03-17 12:53:54+01:00 Europe/Berlin

 - If service has some owner with name "DA_*",
   this service is marked as visible for all owners with name "DA_*"
   in exported data for Netspoc-Web.
 - Command 'print-path' now accepts dual stack objects.

6.078     2025-02-24 14:42:54+01:00 Europe/Berlin

 - New commandline programs export-netvis and print-path have been added.

    These tools output information as JSON to be used in
    a graph visualization.

    export-netvis creates a list of all networks and routers
    of the given topology with information useful for visualizing
    the topology.

    print-path analyzes the given topology and creates a list of
    names of network nodes (routers and networks) between source and
    destination passed via command line.
    To create the list, print-path uses the same path-walk algorithm
    as used in netspoc main program to create ruleset for devices.

 - Aggregate with IPv4 IP 0.0.0.0/0 no longer matches IPv6 addresses.
   Aggregate with IPv6 IP ::/0 no longer matches IPv4 addresses.
 - Fixed split service from auto intf with IPv4, IPv6
 - Command 'export-netspoc' no longer strips comments and ";"
   from description in details of exported service.

6.077     2025-01-27 11:24:26+01:00 Europe/Berlin

 - Allow dual stack service to have protocols icmp and icmpv6 together.
   Generated rules automatically will use icmp for IPv4 and icmpv6 for IPv6.
 - Removed support for separate IPv6 topology in subdirectory 'ipv6/'.
 - Removed options '--ipv6', '-6'.
 - Removed support for NAT of IPv6 objects.
   After changing attributes of IPv6 addresses from 'ip' to 'ip6',
   supporting NAT for IPv6 would need to change NAT attributes
   to 'ip6', 'hidden6', 'identity6' as well.
   This seems to be superflous, since nobody uses NAT for IPv6.
 - Fixed duplicate hosts in print-group from wrong IPv4 network
   in IPv6 aggregate.
 - Fixed panic in print-group from wrong IPv4 network in IPv6 hosts
   generated by 'auto_ipv6_hosts'.
 - Fixed panic in export-netspoc with rule from IPv4 network to
   combined IPv4/v6 auto interface.
 - Changed 'format-netspoc' to print attributes "ip" and "ip6" in
   one line, if object has no more than 3 attributes.

6.076     2025-01-10 09:39:54+01:00 Europe/Berlin

 - Fixed dual stack bridge. IPv4 and IPv6 interfaces were accidently
   processed together.

6.075     2024-12-19 11:53:37+01:00 Europe/Berlin

 - Fix: Attribute 'policy_distribution_point' at dual stack router
   is only used for matching IPv4/IPv6 part of this router.
 - Fixed command 'export-netspoc'.
   Show IPv4 areas for combined zone. Must not accidently overwrite
   IPv4 areas by IPv6 areas, when combined zones have same name.

6.074     2024-12-16 11:52:23+01:00 Europe/Berlin

 - Fixed wrong error on duplicate interface from router in ipv6/ directory.

6.073     2024-12-16 10:16:33+01:00 Europe/Berlin

 - Added support for dual stack objects, having both, IPv4 and IPv6 addresses.
   This simplifies the modeling of a dual stack toplology.
   It is no longer neccessary to model a separate IPv4 and IPv6 topology.
   Rules between dual stack objects will generate ACls for IPv4 and IPv6.
   Rules between dual stack object and pure IPv4 object will silently ignore
   IPv6 address and generate only ACL for IPv4.
   New attributes have been introduced to define dual stack objects:
   - 'ip6' at network, host, interface and aggregate.
   - 'range6' at host.
   - 'unnumbered6' at network and interface.
   - 'negotiated6' at interface.
   Other changes resulting from use of dual stack objects:
   - If dual stack objects are used as border or inclusive border of an area,
     this defines two areas with identical name:
     one in IPv4 topology and one in IPv6 topology.
   - If dual stack objects are used as interfaces of an pathrestriction,
     this also defines two pathrestrictions in IPv4 and IPv6 topology.
     If the second pathrestriction has only one interface or only interfaces
     outside of a loop, it is silently ignored.
   - New attributes 'ipv4_only' and 'ipv6_only' may be used at service or area.
     This will enable only IPv4 or IPv6 part of dual stack objects.
   - The following attributes are applied only to IPv4 part
     if used in dual stack objects:
     'nat', 'bind_nat', 'subnet_of, 'hub', 'spoke'.
   - Command 'export-netspoc' has been changed to show IPv6 address of
     dual stack objects in attribute 'ip6'.
 - Added new attribute 'auto_ipv6_hosts' to automatically generate
   dual stack hosts from pure IPv4 hosts.
   It will generate IPv6 addresses for hosts
   by combining its IPv4 adress with the IPv6 address of its network.
   These attribute values are provided:
   - auto_ipv6_hosts = readable
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1:172:17:1:48
   - auto_ipv6_hosts = binary
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1::ac11:130
   - auto_ipv6_hosts = none
     No IPv6 address is generated
   This attribute can be used at network, area and host.
 - Fixed command 'cut-netspoc':
   Full path to management_instance is marked now.

6.072     2024-09-05 10:22:13+02:00 Europe/Berlin

 - Fixed command 'remove-from-netspoc'.
   A service is no longer removed accidently, if only some rules of
   a service become empty, but other rules remain.
 - Message "swap objects of 'user' and objects of rules" is shown
   even if no attribute multi_owner is given.
 - Fixed check for NAT network being subnet of some other network.
   If multiple networks have dynamic NAT to same address,
   now a warning is given for each network.
 - Fixed check for useless attribute 'subnet_of'
   where supernet has attribute 'has_subnets'.

6.071     2024-07-29 16:47:58+02:00 Europe/Berlin

 - Fixed generated ACL at router R with managed=local for rule where

go/pkg/rename/rename-netspoc.go
package rename

import (
	"fmt"
	"os"
	"strings"

	"github.com/hknutzen/Netspoc/go/pkg/ast"
	"github.com/hknutzen/Netspoc/go/pkg/astset"
	"github.com/hknutzen/Netspoc/go/pkg/oslink"
	"github.com/spf13/pflag"
)

var globalType = map[string]bool{
	"router":          true,
	"network":         true,
	"host":            true,
	"any":             true,
	"group":           true,
	"area":            true,
	"service":         true,
	"owner":           true,
	"protocol":        true,
	"protocolgroup":   true,
	"pathrestriction": true,
	"nat":             true,
	"isakmp":          true,
	"ipsec":           true,
	"crypto":          true,
}

// Interface definition uses network name.
var alias = map[string]string{
	"network": "interface",
}

func getTypeAndName(objName string) (string, string, error) {
	typ, name, ok := strings.Cut(objName, ":")
	if !ok {
		return "", "", fmt.Errorf("Missing type in '%s'", objName)
	}
	return typ, name, nil
}

// Fill subst with mapping from search to replace for given type.
func setupSubst(subst map[string]map[string]string, old, new string) error {
	objType, search, err := getTypeAndName(old)
	if err != nil {
		return err
	}
	newType, replace, err := getTypeAndName(new)
	if err != nil {
		return err
	}
	if objType != newType {
		return fmt.Errorf("Types must be identical in\n - %s\n - %s", old, new)
	}
	if !globalType[objType] {
		return fmt.Errorf("Unknown type %s", objType)
	}
	addSubst := func(typ, search, replace string) error {
		subMap, ok := subst[typ]
		if !ok {
			subMap = make(map[string]string)
			subst[typ] = subMap
		}
		if other := subMap[search]; other != "" {
			return fmt.Errorf("Ambiguous substitution for %s:%s: %s:%s, %s:%s",
				typ, search, typ, other, typ, replace)
		}
		subMap[search] = replace
		return nil
	}

	if err := addSubst(objType, search, replace); err != nil {
		return err
	}

	if other, found := alias[objType]; found {
		addSubst(other, search, replace)
	}
	return nil
}

func process(s *astset.State, subst map[string]map[string]string) {
	s.Modify(func(n ast.Toplevel) bool {
		changed := false
		substitute := func(typ, name string) string {
			if replace, ok := subst[typ][name]; ok {
				changed = true
				return replace
			}
			if typ == "network" || typ == "interface" {
				// Ignore right part of bridged network.
				name, bridged, ok := strings.Cut(name, "/")
				if ok {
					if replace, ok := subst[typ][name]; ok {
						changed = true
						return replace + "/" + bridged
					}
				}
			}
			return name
		}

		var elementList func([]ast.Element)
		var element func(n ast.Element)
		element = func(n ast.Element) {
			typ := n.GetType()
			if typ == "interface" {
				if intf, ok := n.(*ast.IntfRef); ok {
					intf.Router = substitute("router", intf.Router)
					intf.Network = substitute("network", intf.Network)
					return
				}
			}
			switch obj := n.(type) {
			case *ast.NamedRef:
				name := obj.Name
				if typ == "host" && strings.HasPrefix(name, "id:") {
					// ID host is extended by network name: host:id:a.b@c.d.net_name
					parts := strings.Split(name, ".")
					network := parts[len(parts)-1]
					host := strings.Join(parts[:len(parts)-1], ".")
					if replace, ok := subst["host"][host]; ok {
						host = replace
					}
					if replace, ok := subst["network"][network]; ok {
						network = replace
					}
					name = host + "." + network
					if name != obj.Name {
						obj.Name = name
						changed = true
					}
				} else {
					obj.Name = substitute(typ, name)
				}
			case *ast.SimpleAuto:
				elementList(obj.Elements)
			case *ast.AggAuto:
				elementList(obj.Elements)
			case *ast.IntfAuto:
				elementList(obj.Elements)
			case *ast.Intersection:
				elementList(obj.Elements)
			case *ast.Complement:
				element(obj.Element)
			}
		}

		elementList = func(l []ast.Element) {
			for _, n := range l {
				element(n)
			}
		}

		substTypedName := func(v string) string {
			if typ, name, ok := strings.Cut(v, ":"); ok {
				replace := substitute(typ, name)
				return typ + ":" + replace
			}
			return v
		}

		value := func(n *ast.Value) {
			n.Value = substTypedName(n.Value)
		}

		valueList := func(l []*ast.Value) {
			for _, n := range l {
				value(n)
			}
		}

		var attributeList func(l []*ast.Attribute)
		attribute := func(n *ast.Attribute) {
			var m map[string]string
			switch n.Name {
			case "nat_in", "nat_out":
				m = subst["nat"]
			case "owner":
				m = subst["owner"]
			}
			if m != nil {
				for _, v := range n.ValueList {
					if replace, ok := m[v.Value]; ok {
						v.Value = replace
						changed = true
					}
				}
			} else {
				n.Name = substTypedName(n.Name)
			}
			valueList(n.ValueList)
			attributeList(n.ComplexValue)
		}

		attributeList = func(l []*ast.Attribute) {
			for _, n := range l {
				attribute(n)
			}
		}

		namedUnion := func(n *ast.NamedUnion) {
			if n != nil {
				elementList(n.Elements)
			}
		}

		n.SetName(substTypedName(n.GetName()))
		switch x := n.(type) {
		case *ast.TopList:
			elementList(x.Elements)
		case *ast.Protocolgroup:
			valueList(x.ValueList)
		case *ast.TopStruct:
			attributeList(x.Attributes)
		case *ast.Service:
			attributeList(x.Attributes)
			namedUnion(x.User)
			for _, r := range x.Rules {
				namedUnion(r.Src)
				namedUnion(r.Dst)
				valueList(r.Prt.ValueList)
			}
		case *ast.Network:
			attributeList(x.Attributes)
			for _, h := range x.Hosts {
				h.Name = substTypedName(h.Name)
				attributeList(h.ComplexValue)
			}
		case *ast.Router:
			attributeList(x.Attributes)
			for _, intf := range x.Interfaces {
				intf.Name = substTypedName(intf.Name)
				attributeList(intf.ComplexValue)
			}
		case *ast.Area:
			attributeList(x.Attributes)
			namedUnion(x.Border)
			namedUnion(x.InclusiveBorder)
		}
		if changed {
			n.Order()
		}
		return changed
	})
}

func setupPairs(subst map[string]map[string]string, pattern []string) error {
	for len(pattern) > 0 {
		old := pattern[0]
		if len(pattern) < 2 {
			return fmt.Errorf("Missing replace string for '%s'", old)
		}
		new := pattern[1]
		pattern = pattern[2:]
		if err := setupSubst(subst, old, new); err != nil {
			return err
		}
	}
	return nil
}

func readPairs(subst map[string]map[string]string, path string) error {
	bytes, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	pattern := strings.Fields(string(bytes))
	return setupPairs(subst, pattern)
}

func Main(d oslink.Data) int {
	fs := pflag.NewFlagSet(d.Args[0], pflag.ContinueOnError)

	// Setup custom usage function.
	fs.Usage = func() {
		fmt.Fprintf(d.Stderr,
			"Usage: %s [options] FILE|DIR SUBSTITUTION ...\n%s",
			d.Args[0], fs.FlagUsages())
	}

	// Command line flags
	quiet := fs.BoolP("quiet", "q", false, "Don't show number of changes")
	fromFile := fs.StringP("file", "f", "", "Read substitutions from file")
	if err := fs.Parse(d.Args[1:]); err != nil {
		if err == pflag.ErrHelp {
			return 1
		}
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		fs.Usage()
		return 1
	}

	// Argument processing
	args := fs.Args()
	if len(args) == 0 {
		fs.Usage()
		return 1
	}
	path := args[0]

	// Initialize search/replace pairs.
	subst := make(map[string]map[string]string)
	if *fromFile != "" {
		if err := readPairs(subst, *fromFile); err != nil {
			fmt.Fprintf(d.Stderr, "Error: %s\n", err)
			return 1
		}
	}
	if err := setupPairs(subst, args[1:]); err != nil {
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		return 1
	}

	s, err := astset.Read(path)
	if err != nil {
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		return 1
	}
	process(s, subst)
	s.ShowChanged(d.Stderr, *quiet)
	s.Print()
	return 0
}
